### PR TITLE
rustc: Switch `extern` functions to abort by default on panic

### LINKED ISSUE #52652: Abort instead of unwinding past FFI functions
## More updated description (2021-03-11)

With https://github.com/rust-lang/rust/pull/76570 having landed, Rust now aborts when a panic leaves an `extern "C"` function, so the original soundness issue is fixed. However, there is an attribute to opt-out of this behavior, `#[unwind(allowed)]`. Using this attribute is currently unsound, so it should be fixed, removed, or require some form of `unsafe`.

## Updated Description

This is a tracking issue for switching the compiler to abort the program whenever a Rust program panics across an FFI boundary, notably causing any non-Rust-ABI function to panic/unwind.

This is being implemented in #55982 and some regressions were found in the wild so this is also being repurposed as a tracking issue for any breakage that comes up and a place to discuss the breakage. If you're a crate author and you find your way here, don't hesitate to ask questions!

## Original Description

Doing Rust unwinding into non-Rust stack frames is undefined behavior. This was [fixed](https://github.com/rust-lang/rust/pull/46833) in [1.24.0](https://blog.rust-lang.org/2018/02/15/Rust-1.24.html#other-good-stuff), and then reverted (I think by https://github.com/rust-lang/rust/pull/48380 ?) in [1.24.1](https://blog.rust-lang.org/2018/03/01/Rust-1.24.1.html#do-not-abort-when-unwinding-through-ffi) because of a [regression](https://github.com/rust-lang/rust/issues/48251) that affected rlua. 

The latter blog post said:

> The [solution](https://github.com/rust-lang/rust/pull/48572) here is to generate the abort handler, but in a way that `longjmp` won’t trigger it. It’s not 100% clear if this will make it into Rust 1.25; if the landing is smooth, we may backport, otherwise, this functionality will be back in 1.26.

The link PR has landed, but my understanding is that it does *not* change FFI functions back to aborting on unwind (though it looks like it does fix the issue that affected rlua).

This UB is not mentioned in any later release notes.

This issue has been assigned to @BatmanAoD via [this comment](https://github.com/rust-lang/rust/issues/52652#issuecomment-1027361227).

### PR DESCRIPTION
**Pull request by alexcrichton**
This was intended to land way back in 1.24, but it was backed out due to breakage which has long since been fixed. Prior to this PR a crate could panic and unwind past an `extern fn` boundary, but such behavior is UB. For example, this is undefined behavior:

```rust
extern fn foo() {
    panic!();
}

fn main() {
    foo();
}
```

This PR changes the behavior of generated code to be sound-by-default. If an `extern fn` is unwound (panicked through) then it immediately aborts the program. Put another way, no `extern fn` can unwind.

The implementation in this PR is pretty simple because an unstable `#[unwind]` attribute could already be used to tweak the unwinding behavior. As a result this PR basically just switches the default from `#[unwind]` to `#[unwind(abort)]` for all `extern fn` definitions. 

Closes #52652

### CODE REVIEW

**Comment by Mark-Simulacrum** [comment]
FWIW I still personally feel like changing stable behavior without providing a stabilized way to get back the old behavior is not great, but I'm not going to block this myself, both because I think this is fairly rare and the new behavior seems overall later.

**Comment by pietroalbini** [comment]
Maybe we should do a Crater run for this?

**Comment by alexcrichton** [comment]
@bors: r-

Seems fine by me to hold off on a crater run happening!

**Comment by alexcrichton** [comment]
@craterbot run start=master#6b9b97bd9b704f85f0184f7a213cc4d62bd9654c end=try#2e026d6e5d7621634147ad4c8074f1d964dbb1be mode=build-and-test

**Comment by alexcrichton** [comment]
From that report there are four legitimate-looking regressions

* [mujs](https://crater-reports.s3.amazonaws.com/pr-55982/try%232e026d6e5d7621634147ad4c8074f1d964dbb1be/reg/mujs-0.0.3/log.txt) - [faulting code](https://github.com/hean01/mujs-rs/blob/cdfa6bca8855bc53086cf29de956b6b81e2b2272/src/lib.rs#L241-L253)
* [coroutine](https://crater-reports.s3.amazonaws.com/pr-55982/try%232e026d6e5d7621634147ad4c8074f1d964dbb1be/reg/coroutine-0.8.0/log.txt) - [faulting code](https://github.com/rustcc/coroutine-rs/blob/b9ac79eaa190de4fa6d25c83a677c9254c7895f4/src/asymmetric.rs#L152-L161)
* [consistenttime](https://crater-reports.s3.amazonaws.com/pr-55982/try%232e026d6e5d7621634147ad4c8074f1d964dbb1be/reg/consistenttime-0.2.0/log.txt) - [faulting code](https://github.com/valarauca/consistenttime/blob/e489e720c7326eeb78a52e8fbded25a4927c6edf/src/lib.rs#L347-L358)
* [byond](https://crater-reports.s3.amazonaws.com/pr-55982/try%232e026d6e5d7621634147ad4c8074f1d964dbb1be/reg/byond-0.3.2/log.txt) - [this test](https://github.com/PJB3005/byond-rs/blob/7bfbf14f1c9defee7f80eec526cfa8f80b7f4d02/tests/test_calls.rs#L31-L35) causes [this to panic](https://github.com/PJB3005/byond-rs/blob/7bfbf14f1c9defee7f80eec526cfa8f80b7f4d02/src/lib.rs#L51)

In testing I've also noticed that the compiler [no longer emits `nounwind` attributes](https://play.rust-lang.org/?version=stable&mode=debug&edition=2015&gist=c99a4d6b174bf55704b38985b200a546) for `extern` functions. This appears to be a regression in https://github.com/rust-lang/rust/pull/51041, first showing up in Rust 1.28.0.

I've now pushed a follow-up commit to fix the regression, now applying `nounwind` on `extern` function as intended.

**Comment by alexcrichton** [comment]
Do others have thoughts on the crater fallout here?

**Comment by Mark-Simulacrum** [comment]
I'm not quite sure why the mujs case is failing as the `extern` function you link to (https://github.com/hean01/mujs-rs/blob/cdfa6bca8855bc53086cf29de956b6b81e2b2272/src/lib.rs#L241-L253) is not "C" so I'd expect unwinding through it to be fine? Perhaps the cause has been mis-identified? (Also, AFAIK, that "extern fn" does nothing over just "fn", correct?) It looks like maybe this PR is doing too much though -- https://github.com/rust-lang/rust/pull/55982/files#diff-100b8ec08435087bfe75f60887f8aa49 should be a `extern "C" fn` I think, not an `extern fn`...

The coroutine failure looks explicitly designed to unwind across the FFI boundary, which seems a bit weird, since I'm not sure how that actually worked historically. Maybe the C/C++ and Rust unwinding just happens to match up well enough that it works fine, just maybe not running C/C++ destructors?

Similar to the coroutine case, it's interesting to note that the consistenttime library even documents unwinding from the extern C function (https://github.com/valarauca/consistenttime/blob/e489e720c7326eeb78a52e8fbded25a4927c6edf/src/lib.rs#L343-L345). This crate is deprecated though according to the README so I'm not too concerned overall.

The byond-rs test case just seems like it should perhaps be deleted or otherwise restructured -- I doubt the current behavior is actually what the author wanted? 

In summary though I think the only failure that might need more investigation is the mujs case -- the others seem like legitimate failures; we should try and ping authors but I think it's fine to land other than that. The mujs case specifically though needs to be looked into as it seems like maybe this PR is too broad? Specifically, I'd expect `extern "Rust" fn` or `extern fn` to still be allowed to unwind as there's no difference between that and just `fn` AFAIK; this PR I think might add the nounwind LLVM annotation to them too aggressively though.

**Comment by alexcrichton** [comment]
@Mark-Simulacrum oh `extern fn` and `extern "C" fn` [are the same thing](https://play.rust-lang.org/?version=stable&mode=debug&edition=2015&gist=5d40680fcf92eb228208f0e61b9fc0eb)

I do think it'd be good to alert the authors, though!

**Comment by Mark-Simulacrum** [comment]
Must've remembered it the wrong way around.  I guess rustfmt does now transform `extern fn` into extern "C" fn` so that's good.

Beyond that looks like the remaining task here is to get the authors notified -- do you want to take that on? Do you plan to file issues? It might make sense to have a concrete timeline here though (e.g., landing this next Wednesday or something along those lines...)

**Comment by alexcrichton** [comment]
Ok, I'll start the process on this after rustfest starting next week. My plan is to make a tracking issue for this change and cc folks there. I don't think this is terribly urgent to land so we can wait until after December 6 to land during the 1.33 release instead of the 1.32 release to give a maximal amount of time to fix things up.

**Comment by kornelski** [comment]
Is it still going to break the [mozjpeg](https://crates.rs/crates/mozjpeg) crate?

It has IMHO legitimate *and* 100% safe use for unwinding Rust to Rust, but with some safe-to-unwind C code sandwiched in between:

```rust
catch_unwind(|| {
   call_c(rust_callback);
});

extern { fn call_c(*const fn()) }

fn extern "C" rust_callback() {
   panic!();
}
```

**Comment by sdroege** [comment]
> It has IMHO legitimate _and_ 100% safe use for unwinding Rust to Rust

Could it use `extern "Rust"` or similar instead of going via the C ABI?

**Comment by sdroege** [comment]
> I don't see how. The `call_c` bit in my example is actually a call to `libjpeg` which will make more calls to C-ABI functions, which may eventually end with C calling a fatal error callback that is expected to unwind.

I see, so your example above was a bit too simplified. Sorry, should've checked the actual code :) So that's the awkward setjmp/longjmp of libjpeg I assume? I'm not sure that's actually safe but you can work around that with the `#[unwind]` attribute on the function from what I understand.

**Comment by kornelski** [comment]
That is the awkward bit of libjpeg, but to be clear: there is no actual `setjmp` used. I'm trying to unwind Rust-to-Rust with a `panic` instead, so that I can use FFI directly instead of having C wrappers for every C function.

**Comment by kornelski** [comment]
BTW, is the plan to make `#[unwind]` attribute stable at the same time as the aborting behavior? I'd be worried if aborting was stable and `#[unwind]` unstable, as that would break my libjpeg and libpng integrations on stable.

**Comment by alexcrichton** [comment]
As @sdroege mentioned yes @kornelski it's intended that the [example above](https://github.com/rust-lang/rust/pull/55982#issuecomment-444488419) will cause aborts. 

It's currently UB to unwind through C into Rust, we've really put little-to-no-effort to ensure that our unwinding strategy actually works on all platforms when it doesn't go through Rust code, and I don't envision that changing in the near future. The `#[unwind]` attribute was added as a necessary evil when this first came up (but wasn't supposed to be necessary long-term) and is otherwise only tweaking a very low-level detail of LLVM that doesn't relate to the correctness of the API.

For `#[unwind]` to become stable we'd need to provide a guarantee that we actually implement a sound unwinding strategy for all possible platforms to go through C/C++, which I don't really see happening any time soon, especially when we want to leave ourselves to implement unwinding via checked return values on targets where necessary.

**Comment by alexcrichton** [comment]
Looks like Travis never started for the last commit, presumably a missed webhook

**Comment by SimonSapin** [comment]
> but this is currently simply switching rustc's internal default to abort-by-default

This sounds like the default is unchanged, and this PR doesn’t do anything without `#[unwind]`. But that’s not what the test case changes suggest. @alexcrichton Could you clarify the PR description, for the eventual release notes?

**Comment by alexcrichton** [comment]
Sure! @SimonSapin does this clear things up?

---

This was intended to land way back in 1.24, but it was backed out due to breakage which has long since been fixed. Prior to this PR a crate could panic and unwind past an `extern fn` boundary, but such behavior is UB. For example, this is undefined behavior:

```rust
extern fn foo() {
    panic!();
}

fn main() {
    foo();
}
```

This PR changes the behavior of generated code to be sound-by-default. If an `extern fn` is unwound (panicked through) then it immediately aborts the program. Put another way, no `extern fn` can unwind.

The implementation in this PR is pretty simple because an unstable `#[unwind]` attribute could already be used to tweak the unwinding behavior. As a result this PR basically just switches the default from `#[unwind]` to `#[unwind(abort)]` for all `extern fn` definitions.

**Comment by alexcrichton** [comment]
You can explore the UB through the IR of a program like this:

```rust
#![crate_type = "lib"]      
                            
#[inline(never)]            
extern fn foo() {           
    panic!("x");            
}                           
                            
struct A;                   
                            
impl Drop for A {           
    fn drop(&mut self) {    
        extern { fn foo(); }
        unsafe { foo(); }   
    }                       
}                           
                            
pub fn bar() {              
    let _x = A;             
    foo();                  
}                           
```

The compiler places the `nounwind` LLVM attribute on the function `foo`, but it actually ends up unwinding. [According to LLVM](http://llvm.org/docs/LangRef.html):

> If the function does raise an exception, its runtime behavior is undefined

Note that recent compilers have a bug where they *don't* place `nounwind` (fixed in this PR), but 1.25.0 for example placed the `nounwind` attribute.

**Comment by SimonSapin** [comment]
Ok I see. I didn’t know about `nounwind`. Thanks!
